### DISCUSSION TITLE
Kernel Updates with a Capsule

### DISCUSSION THREAD

**Discussion by nwscfox** [OE4T/meta-tegra#1447] [Q&A]
I have yocto with 35.4.1 building uefi capsules that generates both tegra-bl.cap and tegra-kernel.cap
I can copy tegra-bl.cap to esp/EFI/UpgradeCapsule and trigger an update (oe4t-set-uefi-OSIndications), reboot and can see the flash process happen. Once I'm booted, nvbootctrl capsule update status of one and I'm on the "other" slot. 
That all looks perfect.
When I try the same with then tegra-kernel.cap, no flashing seems to happen, capsule update status stays zero. The current slot doesn't change. What am I doing wrong that is preventing the kernel from upgrading?

I would try copying/upgrading both tegra-bl.cap and tegra-kernel.cap at the same time but the esp partition is just barely not big enough to hold them both. Is there an easy way to fix that?

I'm working on an Orin AGX with kirkstone from just a few days ago (2dc5e4)

**Comment by dwalkes** [comment]
I don't have an Orin NX myself, but I know other platforms have limitations on fitting the kernel update in the EFI partition.  I haven't tried to resize it and not sure if that's possible.

If you want to troubleshoot you could capture serial port logs during boot.

Alternatively others have mentioned simply using image copy for devicetree and kernel updates, see [this thread in gitter](https://matrix.to/#/!YBfWVpJwNVtkmqVCPS:gitter.im/$sNItIgUKqmS5SRGdOFwr_Q3jqFuTSUGdsPXxt6KK7rA?via=gitter.im&via=matrix.org&via=3dvisionlabs.com) for instance.

**Reply by mkelly-thegoodpenguin** [reply]
I recently tried altering the size of the EFI partiton to 128Meg for the same reason, so we can do bootloader and kernel at the same time.

I changed the partition size in the custom xml that we're using and then modifed the `images_types_tegra_esp.bbclass` to extract the size from the XML rather than be hard coded.

Seems to have worked okay on an Orin NX,  It entered the capsule update process when triggered to do so.

**Reply by nwscfox** [reply]
Good to know! Were you able to flash the kernel all by itself? What did you do to pick up the kernel portion of the capsule?

**Reply by mkelly-thegoodpenguin** [reply]
So there were two capsules generated in the rootfs in `/opt/nvidia/UpdateCapsule` from the meta-tegra generation process
```
-rw-r--r--    1 root     root      33632006 Mar  9  2018 tegra-bl.cap
-rw-r--r--    1 root     root      43643565 Mar  9  2018 tegra-kernel.cap
```
And I copied over both of them. Previously I had just enough room for the tegra-bl.cap.

I don't think it  has worked with just the kernel on it's own however, so it may not even be doing it when both files are there. It's deleted the file from the `/boot/efi/EFI/UpdateCapsule` but i don't think it processed it.

_-- Edit --_
Debugging shows it does not like the tegra-kernel.cap 
```
FwPackageValidateImageInfo WARNING: image info for 'kernel' has version=13633, expected=0
FwPackageValidateImageInfo WARNING: image info for 'kernel-dtb' has version=13633, expected=0
FwPackageValidateImageInfo WARNING: image info for 'kernel-dtb' has version=13633, expected=0
FwPackageValidateImageInfo WARNING: image info for 'kernel-dtb' has version=13633, expected=0
FwPackageValidateImageInfo WARNING: image info for 'kernel-dtb' has version=13633, expected=0
FwPackageValidateImageInfo WARNING: image info for 'kernel-dtb' has version=13633, expected=0
mb1 not found in package: Not Found
FmpDxe(NVIDIA System Firmware): CheckTheImage() - FmpDeviceLib CheckImage failed. Status = Aborted
FmpDxe(NVIDIA System Firmware): SetTheImage() - Check The Image failed with Aborted.
HandleCapsules: capsule update complete, resetting ...
```
I can't face an more UEFI stuff, so going to go with the same approach as was mentioned in the thread mentioned by [dwalkes](https://github.com/dwalkes) and modify our swupdate to insert the kernel and device tree.

**Reply by nwscfox** [reply]
I ended up insert the kernel/dev tree and that seems workable. In the future, they may make both parts of the capsule work.
